<?xml version="1.0" encoding="utf-8"?>

<launch>
    <!-- Common params -->
    <arg name="camera_ns" default="/stereo_down/left/" />
    <arg name="debug" default="true" />

    <!-- Tracking params -->
    <arg name="detector" default="SIFT" />
    <arg name="descriptor" default="SIFT" />
    <arg name="rep_error" default="3.0" />
    <arg name="inliers" default="35" />
    <arg name="resize" default="0.5" />
    <arg name="target_from_file" default="" />
    <arg name="det_timer" default="15.0" />
    <arg name="use_camera" default="false" />
    <arg name="detect_only" default="false" />
    <arg name="tracker" default="kcf" />
    <arg name="track_thresh" default="0.8" />

    <!-- Visual Servoing params -->
    <arg name="control_freq" default="10.0" />
    <arg name="lambda_x" default="0.35" />
    <arg name="lambda_y" default="0.4" />
    <arg name="lambda_z" default="0.35" />
    <arg name="max_vx" default="0.6" />
    <arg name="max_vy" default="0.5" />
    <arg name="max_wz" default="0.1" />
    <arg name="camera_angle" default="45.0" />
    <arg name="robot_frame" default="sparus2" />
    <arg name="camera_frame" default="left_optical" />
    <arg name="enable_vely" default="false" />

    <!-- Target Selector -->
    <node pkg="merbots_tracking" type="target_selector" name="target_selector" output="screen">
        <remap from="~image" to="$(arg camera_ns)/image_raw" />
        <remap from="~target" to="/det_and_track/target" />
    </node>

    <!-- Detection and Tracking -->
    <node pkg="merbots_tracking" type="det_and_track" name="det_and_track" output="screen">
        <param name="detector" value="$(arg detector)" />
        <param name="descriptor" value="$(arg descriptor)" />
        <param name="debug" value="$(arg debug)" />
        <param name="rep_error" value="$(arg rep_error)" />
        <param name="inliers" value="$(arg inliers)" />
        <param name="resize" value="$(arg resize)" />
        <param name="target_from_file" value="$(arg target_from_file)" />
        <param name="det_timer" value="$(arg det_timer)" />
        <param name="use_camera" value="$(arg use_camera)" />
        <param name="detect_only" value="$(arg detect_only)" />
        <param name="tracker" value="$(arg tracker)" />
        <param name="track_thresh" value="$(arg track_thresh)" />

        <remap from="~image" to="$(arg camera_ns)/image_raw" />
    </node>

    <!-- Visual Servoing -->
    <node pkg="merbots_ibvs" type="ibvs" name="ibvs" output="screen">
        <param name="control_freq" value="$(arg control_freq)" />
        <param name="lambda_x" value="$(arg lambda_x)" />
        <param name="lambda_y" value="$(arg lambda_y)" />
        <param name="lambda_z" value="$(arg lambda_z)" />
        <param name="max_vx" value="$(arg max_vx)" />
        <param name="max_vy" value="$(arg max_vy)" />
        <param name="max_wz" value="$(arg max_wz)" />
        <param name="camera_angle" value="$(arg camera_angle)" />
        <param name="robot_frame" value="$(arg robot_frame)" />
        <param name="camera_frame" value="$(arg camera_frame)" />
        <param name="enable_vely" value="$(arg enable_vely)" />
        <param name="debug" value="$(arg debug)" />

        <remap from="~twist" to="/control/body_velocity_req" />
        <remap from="~target" to="/det_and_track/target" />
        <remap from="~roi" to="/det_and_track/roi" />
        <remap from="~dist" to="/sensors/altitude" />
        <remap from="~image" to="$(arg camera_ns)/image_raw" />
        <remap from="camera_info" to="$(arg camera_ns)/camera_info" />
    </node>

</launch>
